### DISCUSSION TITLE
Code Completion hangs

### DISCUSSION THREAD

**Discussion by cdietrich** [eclipse-langium/langium#1737] [Q&A] [answered]
we sometimes face the situation that the code completion "hangs" forever like this

![Bildschirmfoto 2024-11-04 um 13 19 40](https://github.com/user-attachments/assets/9dd99fba-3be8-42d0-aa14-e1988e945d18)

is it possible the completion is waiting "forever" for the build? e.g waitUntilPhase(Indexed) never finishes?

mixture of manual editing and then invoking content assist

**Comment by msujew** [comment]
@cdietrich I've noticed that as well on certain projects. However, I'm unsure what causes this exactly. I also assume it's some timing issue in the document builder + LSP response handling.

Keep me updated if you have any time to look deeper into this!

**Reply by cdietrich** [reply]
@msujew in the log i did something like

```
    async notifyBuildPhase(documents, state, cancelToken) {
        .....
        if (documents.length === 0) {
            // Don't notify when no document has been processed
            return;
        }
        const listeners = this.buildPhaseListeners.get(state);
        for (const listener of listeners) {
            console.log("inc1")
            await interruptAndCheck(cancelToken);
            console.log("inc2")
            await listener(documents, cancelToken);
            console.log("inc3")
        }
```

and then i see something like

```
inc1
inc2
onBuildPhase: delete listener 5
inc3
```

=> is it possible the callback is disposed before it is executed?

**Reply by cdietrich** [reply]
indeed. bad case

```
notifyBuildPhase file:///xxxx/MyViech.demo.ls 5
inc1
inc2
onBuildPhase: delete listener 5
inc3
```

good case

```
inc1
inc2
waitUntil: bbbbb 5 316
waitUntil: b 5 file:///xxxxMyViech.demo.ls
onBuildPhase: delete listener 5
inc3
```

the bbbbbb is from inside the listener.

=> i wondr what triggers the deleting of the listeners

**Reply by cdietrich** [reply]
hmmm when i split up buildPhaseListeners, onBuildPhase, waitUntil into two lists,
then i can no longer reproduce it.
so i wonder if other services that are cancelled like inlay hint, hover, whatever,
may interfere with the removal?

**Reply by msujew** [reply]
That might be possible. Our disposal logic depends on how functions are compared. I could imagine that there are some false-positives in there, that could accidentally remove a listener when we didn't intend it.

**Reply by cdietrich** [reply]
hmmm am fighting with my bad js skills. should this test work

```
test('bug test', async () => {
        const services = await createServices();
        const workspace = services.shared.workspace;
        const documentFactory = workspace.LangiumDocumentFactory;
        const documents = workspace.LangiumDocuments;
        const uri = URI.parse('file:///test1.txt');
        const document1 = documentFactory.fromString(`
            foo 1 A
            foo 11 B
            bar A
            bar B
        `, uri);
        documents.addDocument(document1);

        const builder = workspace.DocumentBuilder;
        const p1 = builder.waitUntil(DocumentState.IndexedReferences, uri).then(() => {
            console.log('b');
        });
        console.log('p1 created');
        const p2 = builder.waitUntil(DocumentState.IndexedReferences, uri).then(() => {
            console.log('a');
        });
        console.log('p2 created');
        await builder.build([document1], {});
        //await Promise.all([p1, p2]);
        await p2;
        expect(document1.state).toBe(DocumentState.IndexedReferences);
    });
```

**Reply by cdietrich** [reply]
@msujew is it possible we have here the same problem as we have with the classic eclipse listener lists and should do a copy of the list before messing with it

const cpy = [...listeners];
        for (const listener of cpy) {

**Reply by cdietrich** [reply]
(maybe the get should create the copy)

**Reply by cdietrich** [reply]
simplified test

```
    test('xxxxx', async () => {
        const x = new MultiMap<number, string>();
        x.add(1, 'a');
        x.add(1, 'b');
        x.add(1, 'c');
        for (const n of x.get(1)) {
            console.log("n:",n);
            x.delete(1, n);
        }

    });
```

b wont be printed

**Reply by cdietrich** [reply]
https://github.com/eclipse-langium/langium/pull/1738

**Reply by cdietrich** [reply, marked as answer]
https://github.com/eclipse-langium/langium/issues/1740
